### PR TITLE
LGA-157 Extend healthchecks with database and messaging check

### PR DESCRIPTION
**Pull request by rowlando**
## What does this pull request do?

- Use `django-moj-irat` for the existing https://postcodes.io/ healthcheck
- Extends the healthcheck with a "database" check
- Use `django-moj-irat` for the `ping.json`
- Extends the healthcheck with a "messaging" check

See https://github.com/ministryofjustice/django-moj-irat for details.

### All pages are cached by default
`base.py` contains:
```
MIDDLEWARE_CLASSES = (
    'django.middleware.cache.UpdateCacheMiddleware', # <-- LOOK HERE
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'django.middleware.cache.FetchFromCacheMiddleware', # <-- LOOK HERE
)
```
You can read about this here:
https://docs.djangoproject.com/en/1.11/_modules/django/middleware/cache/

This is why we need to purge the cache manually for some pages. Up until this point, `healthcheck.json` and `ping.json` would have been cached for 60 minutes:

```
CACHE_MIDDLEWARE_SECONDS = 3600
```

The import process purges the cache according to [the code in tasks.py](https://github.com/ministryofjustice/laa-legal-adviser-api/blob/2ed57ead6f817600e68ca7783e90adadda840e42/laalaa/apps/advisers/tasks.py#L173):

```
cache.clear()
```

but Igor commented that:

>it could feasibly be because that code runs in a celery task (so different thread/process or even elsewhere entirely) which maybe doesn’t share the same cache as the main app

### Local development convenience
It adds a docker-compose.yml for local dev convenience, which starts the services.

Run:

```
docker-compose run start_services
```

Starting the application takes a while as the db migrations run on start

### CODE REVIEW

File: requirements/base.txt [hunk 1]
```diff
@@ -10,3 +10,6 @@ raven==5.2.0
 requests==2.19.1
 uwsgi==2.0.17.1
 xlrd==0.9.3
+
+#Irat healthcheck and ping package
```
Comment (sldblog) [inline requirements/base.txt]: _nitpick_ Do we need to mention Irat explicitly? It feels we use the healthcheck/ping for our own purposes, not "Incidence Response and Tuning".

File: laalaa/apps/advisers/healthchecks.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    
+    def success_response(self):
+        return HealthcheckResponse(self.name, True)
+
+registry.register_healthcheck(UrlHealthcheck(
+    name='postcodes.io',
+    url='%(base_url)s/postcodes/SW1A1AA' % {
+        'base_url': settings.POSTCODES_IO_URL,
+    },
+    headers={'Content-Type': 'application/json'},
```
Comment (sldblog) [inline laalaa/apps/advisers/healthchecks.py]: 💭 Do we need this header?

File: laalaa/apps/advisers/healthchecks.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        from celery import Celery
+        
+        try:
+            app = Celery('laalaa')
+            stats = app.control.inspect().stats()
+            if not stats:
+                return self.error_response('No running workers were found.')
+            workers = stats.values()
+            if len(workers) == 0:
+                return self.error_response('No workers running')
```
Comment (sldblog) [inline laalaa/apps/advisers/healthchecks.py]: _consistency nitpick_
The rest of the sentences have full stops at the end :neckbeard:

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+version: '3.2'
+services:
+  start_services:
```
Comment (sldblog) [inline docker-compose.yml]: Could we document the usage of this in the README as well, please?

File: laalaa/apps/advisers/healthchecks.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    def __call__(self, *args, **kwargs):
+        from celery import Celery
+        
+        try:
+            app = Celery('laalaa')
+            stats = app.control.inspect().stats()
+            if not stats:
+                return self.error_response('No running workers were found.')
+            workers = stats.values()
+            if len(workers) == 0:
```
Comment (sldblog) [inline laalaa/apps/advisers/healthchecks.py]: 🙋‍♂️ It seems `len(workers) == 0` and `not workers` are equivalent in Python:

```
>>> list = []
>>> len(list) == 0
True
>>> not list
True

>>> list = [1]
>>> len(list) == 0
False
>>> not list
False
```

File: laalaa/settings/base.py [hunk 6]
```diff
@@ -19,13 +19,16 @@
...
+AUTODISCOVER_HEALTHCHECKS = True
+
 # Quick-start development settings - unsuitable for production
 # See https://docs.djangoproject.com/en/1.7/howto/deployment/checklist/
 
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = os.environ.get('SECRET_KEY', 'DEV_KEY')
 
-DEBUG = False
+DEBUG = True
```
[the remark this answers is not in the document]
Comment (rowlando) [reply laalaa/settings/base.py]: Good idea!

File: laalaa/settings/base.py [hunk 7]
```diff
@@ -19,13 +19,16 @@
 
 sys.path.insert(0, root('apps'))
 
+HEALTHCHECKS = ['moj_irat.healthchecks.database_healthcheck']
+AUTODISCOVER_HEALTHCHECKS = True
```
[the remark this answers is not in the document]
Comment (rowlando) [reply laalaa/settings/base.py]: Added

File: laalaa/settings/base.py [hunk 8]
```diff
@@ -87,14 +90,7 @@
...
 DATABASES = {
-    'default': {
-        'ENGINE': 'django.contrib.gis.db.backends.postgis',
-        'NAME': os.environ.get('DB_USERNAME', 'laalaa'),
-        'USER': os.environ.get('DB_USERNAME', 'postgres'),
-        'PASSWORD': os.environ.get('DB_PASSWORD', ''),
-        'HOST': os.environ.get('DB_HOST', '127.0.0.1'),
-        'PORT': os.environ.get('DB_PORT', '5432'),
-    }
+    'default': { 'ENGINE': 'django.contrib.gis.db.backends.postgis', 'NAME': os.environ.get('DB_USERNAME', 'laalaa'), 'USER': os.environ.get('DB_USERNAME', 'postgres'), 'PASSWORD': os.environ.get('DB_PASSWORD', ''), 'HOST': os.environ.get('DB_HOST', '127.0.0.1'), 'PORT': os.environ.get('DB_PORT', '5432'), }
```
[the remark this answers is not in the document]
Comment (rowlando) [reply laalaa/settings/base.py]: No - good catch.

File: laalaa/apps/advisers/healthchecks.py (hunk 2 above)
Comment (rowlando) [reply laalaa/apps/advisers/healthchecks.py]: Nope.

File: docker-compose.yml (hunk 4 above)
Comment (rowlando) [reply docker-compose.yml]: Good point!

File: requirements/base.txt (hunk 1 above)
Comment (rowlando) [reply requirements/base.txt]: Nope. Removed.

File: laalaa/apps/advisers/healthchecks.py (hunk 3 above)
Comment (rowlando) [reply laalaa/apps/advisers/healthchecks.py]: Fixed

File: laalaa/apps/advisers/healthchecks.py (hunk 5 above)
Comment (rowlando) [reply laalaa/apps/advisers/healthchecks.py]: Nice!

File: laalaa/settings/local.py.example [hunk 9]
```diff
@@ -0,0 +1,4 @@
+from .common import *
```
Comment (sldblog) [inline laalaa/settings/local.py.example]: This fails for me with `No module named common` :(

Comment (sldblog) [reply laalaa/settings/local.py.example]: It seems we can remove the import and the `TEMPLATE_DEBUG` too:

`?: (1_8.W001) The standalone TEMPLATE_* settings were deprecated in Django 1.8 and the TEMPLATES dictionary takes precedence. You must put the values of the following settings into your default TEMPLATES dict: TEMPLATE_DEBUG.`.

Comment (rowlando) [reply laalaa/settings/local.py.example]: Agreed. Good spot.

File: laalaa/apps/advisers/healthchecks.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+class CeleryWorkersHealthcheck(object):
+    def __init__(self, name):
+        self.name = name
+    
+    def __call__(self, *args, **kwargs):
+        from celery import Celery
+        
+        try:
+            app = Celery('laalaa')
+            stats = app.control.inspect().stats()
```
Comment (sldblog) [inline laalaa/apps/advisers/healthchecks.py]: I wonder if we could add a test for this check? Maybe port it back to the irat repo?

Comment (rowlando) [reply laalaa/apps/advisers/healthchecks.py]: Yeap, once it's merged, I'll ping the author of the irat repo about porting it back and helping with a test.

**Pull request merged**